src/libpmemfile-posix/callbacks.c

```diff
@@ -0,0 +1,243 @@
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * callbacks.c -- transaction callback subsystem
+ */
+
+#include "callbacks.h"
+#include "internal.h"
+#include "os_thread.h"
+#include "out.h"
+#include "util.h"
+
+#include <errno.h>
+
+struct tx_callback {
+	cb_basic func;
+	void *arg;
+};
+
+struct tx_callback_array {
+	struct tx_callback *arr;
+
+	/* Size of callbacks array. */
+	unsigned size;
+
+	/* Number of registered callbacks. */
+	unsigned used;
+};
+
+struct all_callbacks {
+	struct tx_callback_array forward;
+	struct tx_callback_array backward;
+};
+
+static os_tls_key_t callbacks_key;
+
+/*
+ * cb_get -- returns current per-thread callback configuration
+ */
+static struct all_callbacks *
+cb_get(void)
+{
+	struct all_callbacks *c = os_tls_get(callbacks_key);
+	if (!c) {
+		c = calloc(1, sizeof(struct all_callbacks) * MAX_TX_STAGE);
+		int ret = os_tls_set(callbacks_key, c);
+		if (ret) {
+			errno = ret;
+			FATAL("!os_tls_set");
+		}
+	}
+
+	return c;
+}
+
+/*
+ * cb_check -- (internal) check whether current state
+ * allows changing transaction callbacks
+ */
+static void
+cb_check(const char *func)
+{
+	if (pmemobj_tx_stage() == TX_STAGE_NONE)
+		FATAL("%s called outside of transaction", func);
+}
+
+/*
+ * cb_append -- (internal) appends callback to specified array
+ */
+static int
+cb_append(struct tx_callback_array *cb, cb_basic func, void *arg)
+{
+	if (cb->used == cb->size) {
+		unsigned count = cb->size * 2;
+		if (count == 0)
+			count = 4;
+
+		void *new_arr = realloc(cb->arr, count * sizeof(cb->arr[0]));
```

Comment:
It's still allocated, but we don't have to free it. It will be used by the next transaction and freed on thread exit.